classes/observers/listeners/PKPUsageEventLogListener.inc.php

```diff
@@ -0,0 +1,305 @@
+<?php
+
+/**
+ * @file classes/observers/listeners/PKPUsageEventLogListener.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class PKPUsageEventLogListener
+ * @ingroup observers_listeners
+ *
+ * @brief Listener listening for the usage events.
+ */
+
+namespace PKP\observers\listeners;
+
+use APP\core\Application;
+use APP\facades\Repo;
+use APP\observers\events\UsageEvent;
+use APP\statistics\StatisticsHelper;
+use GeoIp2\Database\Reader;
+use PKP\cache\CacheManager;
+use PKP\cache\FileCache;
+use PKP\file\PrivateFileManager;
+use Sokil\IsoCodes\IsoCodesFactory;
+
+class PKPUsageEventLogListener
+{
+    public FileCache $geoDataCache;
+
+    /**
+     * Handle the event.
+     */
+    public function handle(UsageEvent $usageEvent): void
+    {
+        $usageEventArray = $this->prepareUsageEvent($usageEvent);
+        $this->logUsageEvent($usageEventArray);
+    }
+
+    /**
+     * Log the usage event
+     */
+    public function logUsageEvent(array $usageEventArray): void
+    {
+        $usageEventLogEntry = json_encode($usageEventArray) . PHP_EOL;
+
+        // Log file name (from the current day)
+        $logFileName = $this->getUsageEventLogFileName();
+
+        // Write the event to the log file
+        // Keep the locking in order not to care about the filesystems's block sizes and if the file is on a local filesystem
+        $fp = fopen($logFileName, 'a+b');
+        if (flock($fp, LOCK_EX)) {
+            fwrite($fp, $usageEventLogEntry);
+            flock($fp, LOCK_UN);
+        } else {
+            // Couldn't lock the file.
+            assert(false);
```

Comment:
Hmmm... I think we do not need to stop the propagation of an event to other listeners -- we do not know what other listeners want to do with the event...
I do not know what to do here with assert, maybe I could just leave it out or throw an exception?
@asmecher, what do you think?

Reply:
I don't think we want a `fatalError` in this case -- that would cause the reader to encounter a 500. For the moment `error_log` is probably the best place to stash a message like this.

Reply:
I changed all asserts in this file to error_log.